src/kafka-net/KafkaTcpSocket.cs

```diff
@@ -179,23 +192,36 @@ private void ProcessNetworkstreamTasks(NetworkStream netStream)
             //https://msdn.microsoft.com/en-us/library/z2xae4f4.aspx
             while (_disposeToken.IsCancellationRequested == false && netStream != null)
             {
-                Task sendDataReady = Task.WhenAll(writeTask, _sendTaskQueue.OnHasDataAvailable(_disposeToken.Token));
-                Task readDataReady = Task.WhenAll(readTask, _readTaskQueue.OnHasDataAvailable(_disposeToken.Token));
-
-                Task.WaitAny(sendDataReady, readDataReady);
-
-                var exception = new[] { writeTask, readTask }
-                    .Where(x => x.IsFaulted && x.Exception != null)
-                    .SelectMany(x => x.Exception.InnerExceptions)
-                    .FirstOrDefault();
+                Task sendDataReady = Task.WhenAll(_sendTaskQueue.OnHasDataAvailable(_disposeToken.Token), lastWriteTask);
+                Task readDataReady = Task.WhenAll(_readTaskQueue.OnHasDataAvailable(_disposeToken.Token), lastReadTask);
+                await Task.WhenAny(sendDataReady, readDataReady, _disposeTask).ConfigureAwait(false);
+                if (_disposeToken.IsCancellationRequested) return;
 
+                var exception = FindExceptionInLastReadOrWriteTasks(lastWriteTask, readDataReady);
```

Comment:
I'm a bit confused, shouldn't it be lastWriteTask and lastReadTask? or sendDataReady and readDataReady? but I don't understand why lastWriteTask and readDataReady